dali/pipeline/data/tensor_list.h

```diff
@@ -116,15 +124,19 @@ class DLL_PUBLIC TensorList : public Buffer<Backend> {
    *
    * When this function is called, the calling object shares the
    * underlying allocation of the input TensorList. Its size, type
-   * and shape are set to match the calling TensorList. While this 
-   * list shares data with another list, 'shares_data()' will 
+   * and shape are set to match the calling TensorList. While this
+   * list shares data with another list, 'shares_data()' will
    * return 'true'.
    */
   DLL_PUBLIC inline void ShareData(TensorList<Backend> *other) {
     DALI_ENFORCE(other != nullptr, "Input TensorList is nullptr");
     DALI_ENFORCE(IsValidType(other->type_), "To share data, "
         "the input TensorList must have a valid data type");
 
+    // Tensor view of this TensorList is no longer valid
+    delete tensor_view_;
+    tensor_view_ = nullptr;
```

Comment:
I don't know if python will handle this, what if we call AsTensor, return pointer, call ShareData, then pointer we just returned to python is no longer valid but python still can reference it, or I'm wrong?

Reply:
This is a much longer discussion - in order to handle this we would need to store information about all the objects sharing data with a particular tensor or tensorlist and notify them that the data is no longer valid. This would potentially have performance implications.

Reply:
Oh, I see what you mean. I was referring to the other problem with sharing data. Hmmm, instead of deleting the pointer I guess I could keep it but invoke `ShareData` again after the internal pointer to `data_` changes.

Reply:
It is some solution to our problem.
Regarding cores reference I following approaches
- an object which shares its own data could keep a list of reference and when change notify and invalidate the state of other objects
- an object that shares data of other object checks its validity
- some kind of RCU mechanism - then we will keep all data valid
First and second requires a lot of synchronization that could hit performance.
Another solution is to leave it as design assumption - like iterator invalidation when data we are iterating over changes.